ENH: support pathlib in ReaderBase

tylerjereddy opened the pull request:
* the creation of a `Universe` object involving GROMACS formats should now accept `pathlib`
objects as requested in gh-2497; resolving that issue entirely is a separate matter, this branch focuses only on the GROMACS format and only if you proceed via the public `Universe` creation API

* probably could make it work in Cython directly, though I'm not convinced we need to do that for this part of the fix at least, and we probably should still support strings as well of course

tylerjereddy committed:
ENH: support GROMACS pathlib

* the creation of a `Universe` object involving
GROMACS formats should now accept `pathlib`
objects as requested in gh-2497; resolving that
issue entirely is a separate matter, this branch
focuses only on the GROMACS format and only if
you proceed via the public `Universe` creation
API

* probably could make it work in Cython directly,
though I'm not convinced we need to do that for
this part of the fix at least, and we probably
should still support strings as well of course

IAlibay reviewed (commented):
Shouldn't this be a ReaderBase thing instead to store self.filename properly?

hmacdope commented:
> Shouldn't this be a ReaderBase thing instead to store self.filename properly?

Seconding this, can this not be dealt with at `ReaderBase` level to fix it for all readers?

tylerjereddy committed:
MAINT: PR 3935 revisions

* move the handling of `pathlib` objects
upstream to `ReaderBase` based on reviewer
feedback

* add one more format test related to this

tylerjereddy commented:
Ok, I moved the patch to `ReaderBase` and added a test for one more format (`DCD`) as a sniff test for generalization.

I think that's probably good enough for now, but note that:

1) This still doesn't cover all readers because `SingleFrameReaderBase` has its own `self.filename` -- I'd be happy to delay patching/testing that for now.
2) The testing could probably be centralized/parametrized somewhere, covering a bunch of formats rather than testing the same pass-through for just two formats. Maybe good enough for now though..

IAlibay reviewed (approved):
Assuming this passes tests I'm happy for this PR to go through. @tylerjereddy could you raise an issue about the missing bits? Maybe we should move stuff to ProtoReader in the future but it seems fine for this current issue?

tylerjereddy commented:
Eh, the failure is related--this is the problem with inheritance... I'd just as soon scope this back to GROMACS and move on.

I'll see if I can add a shim to fix it though.

tylerjereddy committed:
MAINT: PR 3935 revisions

* `NamedStream` gets special treatment (no string coercion)
in `ReaderBase`

tylerjereddy commented:
Ok, I pushed in a commit to shim around our `NamedStream` stuff in the base class, and this seems to allow the full suite to pass for me locally at least.

gh-3937 contains the the follow-up reminders I noted above.

hmacdope commented:
LGTM @tylerjereddy

tylerjereddy commented:
Ok, I have approvals from 3 core developers and CI is passing, in it goes.
